MovieNight/MovieMatcher.swift

```diff
@@ -0,0 +1,150 @@
+                    for credits in moviesCredits {
+                        let movieCreditsMatchedByActor = credits.filter { actors.contains($0.actor) }
+                        
+                        // Ger from allRecommendedMovies only the movies with 'actors matched'
+                        for credit in movieCreditsMatchedByActor {
+                            let moviesMatched = self.allRecommendedMovies.filter { $0.id == credit.movieId }
+                            for movie in moviesMatched {
+                                self.moviesMatchedByActor.append(movie)
+                            }
+                        }
+                    }
+                    
+                    if !(self.moviesMatchedByGenre.isEmpty) {
+                        for movie in self.moviesMatchedByGenre {
+                            self.matchedMoviesByGenreAndActor.append(movie)
+                        }
+                    }
+                    
+                    if !(self.moviesMatchedByActor.isEmpty) {
+                        for movie in self.moviesMatchedByActor {
+                            self.matchedMoviesByGenreAndActor.append(movie)
+                        }
+                    }
+                    
+                    // Can be called to show the results only after all the credits were checked
+                    DispatchQueue.main.async {
+                        if !(self.matchedMoviesByGenreAndActor.isEmpty) {
+                            completionHandler(self.matchedMoviesByGenreAndActor.uniq())
+                            print(self.matchedMoviesByGenreAndActor.count)
+                            
+                        } else {
+                            completionHandler(self.allRecommendedMovies.uniq())
+                            print(self.allRecommendedMovies.count)
+                        }
+                    }
+                })
+            }
+        })
+    }
+    
+    private func getMovieRecommendations(from movies: [Movie], completionHandler: @escaping (_ allRecommendedMovies: [[Movie]])-> Void) {
+        let group = DispatchGroup()
+        let moviesId = movies.map { String($0.id) }
+        let endpoints = moviesId.map { MovieNightEndpoint.MovieRecommendations(id: $0) }
+        
+        for endpoint in endpoints {
+            group.enter()
+            apiCLient.fetchMovies(endpoint: endpoint, completion: { (result) in
+                switch result {
+                    case .failure(let error):
+                        print(error)
+                    case .success(let resource , _):
+                        self.recommendedMovies.append(resource)
+                }
+                group.leave()
+            })
+        }
+        
+        group.notify(queue: DispatchQueue.global(qos: .userInitiated)) {
+            print("All async calls were run!")
+            completionHandler(self.recommendedMovies)
```

Comment:
I think you could flatten self.recommendedMovies here to simplify the calling code and also call uniq (as we discussed, to avoid wasting resources)